### PR TITLE
Fixed Upload category image with same image name

### PR DESCRIPTION
**Pull request by ravi-chandra3197**
### Description (*)

Fixed Upload category image with same image name
### Fixed Issues (if relevant)

1. magento/magento2#17661: Upload category image with same image name

### Manual testing scenarios (*)

1. Take a category.
2. Upload Category image.
3. Save your category.
4. Take a another Category and upload image with another image but same image name.
5. Save your category.

### Contribution checklist (*)
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [x] All new or changed code is covered with unit/integration tests (if applicable)
 - [x] All automated tests passed successfully (all builds on Travis CI are green)

### CODE REVIEW

**Comment by magento-engcom-team** [comment]
Hi @ravi-chandra3197. Thank you for your contribution 
Here is some useful tips how you can test your changes using Magento test environment.
Add the comment under your pull request to deploy test or vanilla Magento instance:
- `@magento-engcom-team give me test instance` - deploy test instance based on PR changes
- `@magento-engcom-team give me 2.3-develop instance` - deploy vanilla Magento instance

For more details, please, review the [Magento Contributor Assistant](https://devdocs.magento.com/guides/v2.2/contributor-guide/contributing.html#contributor-assist) documentation

File: lib/internal/Magento/Framework/File/Uploader.php [hunk 1]
```diff
@@ -207,6 +207,10 @@ public function save($destinationFolder, $newFileName = null)
         $this->_result = false;
         $destinationFile = $destinationFolder;
         $fileName = isset($newFileName) ? $newFileName : $this->_file['name'];
+        $basePath = str_replace('/tmp/','/',$destinationFolder);
+        if (file_exists($basePath.'/'.$this->_file['name'])) {
+            $fileName = pathinfo($this->_file['name'], PATHINFO_FILENAME)."_1.".pathinfo($this->_file['name'], PATHINFO_EXTENSION);
```
Comment (orlangur) [inline lib/internal/Magento/Framework/File/Uploader.php]: What if `_1` already exists? Also, please get rid of `pathinfo` calls.

**Review by orlangur — CHANGES_REQUESTED** [review]

File: lib/internal/Magento/Framework/File/Uploader.php (hunk 1 above)
Comment (ravi-chandra3197) [reply lib/internal/Magento/Framework/File/Uploader.php]: Hello @orlangur
_1 append in the file name for rename it.
Any suggestion?

**Comment by sivaschenko** [comment]
Closing this pull request due to inactivity, feel free to reopen in case you'd like to continue progress on it.

**Comment by ravi-chandra3197** [comment]
Hello @sivaschenko
I this fix working for me. I have tested it if the category image exists with the same name so we have just append  _1 at end of the file name.

**Comment by sivaschenko** [comment]
Hi @ravi-chandra3197 as it was mentioned by @orlangur there is a potential problem in case the file with name `fileName_1.fileExtension` already exists as well

**Comment by ravi-chandra3197** [comment]
Hello @sivaschenko 
if `fileName_1.fileExtension` already exists so it append _1 for unique image name.

**Comment by sivaschenko** [comment]
@ravi-chandra3197 here is the situation that concerns me:

*Steps:*
There are two files in the system: `fileName.jpg` and `fileName_1.jpg`.
A new file is uploaded: `fileName.jpg`

Actual Result:
The solution provided in this pull request verifies that `fileName.jpg` does exist in the system, appends `_1` to the file name, and saves the file as `fileName_1.jpg` overriding the existing `fileName_1.jpg`

**Comment by ravi-chandra3197** [comment]
Hello @sivaschenko 
I have updated code can you review it.
Now it resolves the issue 

Steps:
There are two files in the system: fileName.jpg and fileName_1.jpg.
A new file is uploaded: fileName.jpg

Actual Result:
The solution provided in this pull request verifies that fileName.jpg does exist in the system, appends _1 to the file name, and saves the file as fileName_1.jpg

**Comment by tnsezer** [comment]
@ravi-chandra3197 

> Hello @sivaschenko
> I have updated code can you review it.
> Now it resolves the issue
> 
> Steps:
> There are two files in the system: fileName.jpg and fileName_1.jpg.
> A new file is uploaded: fileName.jpg
> 
> Actual Result:
> The solution provided in this pull request verifies that fileName.jpg does exist in the system, appends _1 to the file name, and saves the file as fileName_1.jpg

Why you don't add a counter and merge it with the counter? Also you can add a limit and if the counter react the limit you can replace the fileName with any unique name.

```
$tmpFileName = $this->_file['name'];
$i = 1;
while(true) {
            $tmpFileNameNew = pathinfo($tmpFileName, PATHINFO_FILENAME)."_".$i.".".pathinfo($tmpFileName, PATHINFO_EXTENSION);
if(!file_exists($basePath.'/'.$tmpFileNameNew)){
$tmpFileName = $tmpFileNameNew;
break;
}
$i++;
        }
```

File: lib/internal/Magento/Framework/File/Uploader.php [hunk 2]
```diff
@@ -207,6 +207,12 @@ public function save($destinationFolder, $newFileName = null)
         $this->_result = false;
         $destinationFile = $destinationFolder;
         $fileName = isset($newFileName) ? $newFileName : $this->_file['name'];
+        $basePath = str_replace('/tmp/','/',$destinationFolder);
+        $tmpFileName = $this->_file['name'];
+        while(file_exists($basePath.'/'.$tmpFileName)) {
+            $tmpFileName = pathinfo($tmpFileName, PATHINFO_FILENAME)."_1.".pathinfo($tmpFileName, PATHINFO_EXTENSION);
+        }
+        $fileName = $tmpFileName;
```
Comment (orlangur) [inline lib/internal/Magento/Framework/File/Uploader.php]: There is already such logic in this file: https://github.com/magento/magento2/blob/62d6e69276acfa047ac7e3782a53d52183e4374c/lib/internal/Magento/Framework/File/Uploader.php#L226

Please fix properly.

**Review by orlangur — CHANGES_REQUESTED** [review]

File: lib/internal/Magento/Framework/File/Uploader.php (hunk 2 above)
Comment (ravi-chandra3197) [reply lib/internal/Magento/Framework/File/Uploader.php]: Hello @orlangur 
I have updated code can you review it.

File: lib/internal/Magento/Framework/File/Uploader.php [hunk 3]
```diff
@@ -604,10 +604,12 @@ private function _createDestinationFolder($destinationFolder)
...
     {
         $fileInfo = pathinfo($destinationFile);
-        if (file_exists($destinationFile)) {
+        $destinationBaseFile = str_replace('/tmp/','/',$destinationFile);
+        if (file_exists($destinationFile) || file_exists($destinationBaseFile)) {
             $index = 1;
             $baseName = $fileInfo['filename'] . '.' . $fileInfo['extension'];
-            while (file_exists($fileInfo['dirname'] . '/' . $baseName)) {
+            $dirBasename = str_replace('/tmp/','/',$fileInfo['dirname']);
+            while (file_exists($fileInfo['dirname'] . '/' . $baseName) || file_exists($dirBasename . '/' . $baseName)) {
```
Comment (orlangur) [inline lib/internal/Magento/Framework/File/Uploader.php:612]: Why previous logic does not find a proper unique file name?

**Review by orlangur — CHANGES_REQUESTED** [review]

File: lib/internal/Magento/Framework/File/Uploader.php (hunk 3 above)
Comment (ravi-chandra3197) [reply lib/internal/Magento/Framework/File/Uploader.php:612]: each time we have to check file name in both destination folders like for category image we have check following destination.
pub/media/catalog/category
pub/media/catalog/tmp/category
in default it check only pub/media/catalog/tmp/category but if same file exist in pub/media/catalog/category so it replace it.

Comment (ravi-chandra3197) [reply lib/internal/Magento/Framework/File/Uploader.php:612]: Hello @orlangur

https://github.com/magento/magento2/pull/20763#issuecomment-469679520
It also check for above situation

**Comment by ravi-chandra3197** [comment]
Hello @orlangur
I have updated this PR can you review it.

**Comment by ravi-chandra3197** [comment]
Hello @orlangur
I have updated this PR can you review it.

**Comment by ravi-chandra3197** [comment]
Hello @magento-engcom-team , @orlangur
Any update for this PR.

File: lib/internal/Magento/Framework/File/Uploader.php [hunk 4]
```diff
@@ -604,10 +604,12 @@ private function _createDestinationFolder($destinationFolder)
     public static function getNewFileName($destinationFile)
     {
         $fileInfo = pathinfo($destinationFile);
-        if (file_exists($destinationFile)) {
+        $destinationBaseFile = str_replace('/tmp/','/',$destinationFile);
```
Comment (sivaschenko) [inline lib/internal/Magento/Framework/File/Uploader.php:607]: This function should be abstract and should not rely on the hardcoded path samples like `/tmp/`

**Review by sivaschenko — CHANGES_REQUESTED** [review]
Thanks for updates @ravi-chandra3197 . I don't think the abstract `getNewFileName` function should be changed.

 - If the issue can be reproduced only for category image - I'd place the fix in the Category module
 - If it's the global issue, I think it might be caused by incorrect `$destinationFile` argument passed to this method.

In both cases, the issue should be fixed higher in the stack in my opinion

File: lib/internal/Magento/Framework/File/Uploader.php (hunk 4 above)
Comment (orlangur) [reply lib/internal/Magento/Framework/File/Uploader.php:607]: @sivaschenko please check other comments when doing review

Comment (sivaschenko) [reply lib/internal/Magento/Framework/File/Uploader.php:607]: @orlangur which comment are you referring to?

Comment (orlangur) [reply lib/internal/Magento/Framework/File/Uploader.php:607]: https://github.com/magento/magento2/pull/20763#discussion_r273345620 answers your question

Comment (sivaschenko) [reply lib/internal/Magento/Framework/File/Uploader.php:607]: I've seen that comment and the problems mentioned in my review are still actual

**Comment by sidolov** [comment]
@ravi-chandra3197 , I am closing this PR now due to inactivity.
Please reopen and update if you wish to continue.
Thank you for the collaboration!

**Pull request closed without merging**